internal/overlord/logstate/manager.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+type LogManager interface {
+	PlanChanged(*plan.Plan)
+	NewService(serviceName string, buffer *servicelog.RingBuffer)
+}
+
+type logManager struct {
+	// Store services and selected log targets, as defined by plan
+	planTargets serviceTargetSet
+	// Store buffers for each running service
+	buffers map[string]*servicelog.RingBuffer
+	// Store pumpers for each service & target
+	pumpers map[serviceTarget]*pumper
+	// Pumpers to be added once the service starts
+	toAdd serviceTargetSet
+}
+
+func NewLogManager() LogManager {
+	return &logManager{
+		buffers: map[string]*servicelog.RingBuffer{},
+		pumpers: map[serviceTarget]*pumper{},
+		toAdd:   serviceTargetSet{},
+	}
+}
+
+func (m *logManager) PlanChanged(p *plan.Plan) {
+	// Update planTargets
+	m.planTargets = serviceTargetSet{}
+	for serviceName, service := range p.Services {
+		for _, targetName := range service.LogTargets {
+			target, ok := p.LogTargets[targetName]
+			if ok {
+				m.planTargets.Add(serviceTarget{
+					serviceName,
+					*target,
+				})
+			} else {
+				logger.Noticef("log manager: target %q doesn't exist", targetName)
+			}
+		}
+	}
+
+	add, remove := deltas(m.planTargets, m.pumpers)
+
+	// Remove obsolete pumpers
+	for st := range remove {
+		m.pumpers[st].Stop()
+		delete(m.pumpers, st)
+	}
+
+	// Add new pumpers
+	m.toAdd.AddAll(add)
+	m.addNewPumpers()
+}
+
+// Register the service (and buffer) with the log manager, and create new
+// pumpers that are on the manager's to-do list.
+func (m *logManager) NewService(serviceName string, buffer *servicelog.RingBuffer) {
+	fmt.Println("new service started")
```

Comment:
I usually put a TODO comment on this kind of debugging, so it's easier to spot in case I accidentally leave it in.